databricks/koalas/indexes.py

```diff
@@ -399,6 +400,70 @@ def copy(self, name=None):
             result.name = name
         return result
 
+    def symmetric_difference(self, other, result_name=None, sort=None):
+        """
+        Compute the symmetric difference of two Index objects.
+
+        Parameters
+        ----------
+        other : Index or array-like
+        result_name : str
+        sort : True or None, default None
+            Whether to sort the resulting index.
+            * True : Attempt to sort the result.
+            * None : Do not sort the result.
+
+        Returns
+        -------
+        symmetric_difference : Index
+
+        Notes
+        -----
+        ``symmetric_difference`` contains elements that appear in either
+        ``idx1`` or ``idx2`` but not both. Equivalent to the Index created by
+        ``idx1.difference(idx2) | idx2.difference(idx1)`` with duplicates
+        dropped.
+
+        Examples
+        --------
+        >>> s1 = ks.Series([1, 2, 3, 4], index=[1, 2, 3, 4])
+        >>> s2 = ks.Series([1, 2, 3, 4], index=[2, 3, 4, 5])
+
+        >>> s1.index.symmetric_difference(s2.index)
+        Int64Index([5, 1], dtype='int64')
+
+        You can set name of result Index.
+
+        >>> s1.index.symmetric_difference(s2.index, result_name='koalas')
+        Int64Index([5, 1], dtype='int64', name='koalas')
+
+        You can set sort to `True`, if you want to sort the resulting index.
+
+        >>> s1.index.symmetric_difference(s2.index, sort=True)
+        Int64Index([1, 5], dtype='int64')
+
+        You can also use the ``^`` operator:
+
+        >>> s1.index ^ s2.index
+        Int64Index([5, 1], dtype='int64')
+        """
+        sdf_self = self._kdf._sdf.select(self._internal.index_scols)
```

Comment:
We need to check the index levels and data types here? E.g., what if `Index ^ MultiIndex`?
Could you add such tests?

Reply:
Sure, i'll add

BTW, i found some funny behavior in pandas, not pretty sure, maybe a bug in pandas.

There is an index and a MultiIndex of pandas here.

```python
>>> idx = pd.Index([1, 2, 3, 4])
>>> midx = pd.MultiIndex([['lama', 'cow', 'falcon'],
...                        ['speed', 'weight', 'length']],
...                       [[0, 0, 0, 1, 1, 1, 2, 2, 2],
...                        [0, 0, 0, 0, 1, 2, 0, 1, 2]])
```

then, below seems to work pretty well,

```python
>>> idx ^ midx
Index([                   1,                    2,                    3,
                          4,    ('lama', 'speed'),     ('cow', 'speed'),
          ('cow', 'weight'),    ('cow', 'length'),  ('falcon', 'speed'),
       ('falcon', 'weight'), ('falcon', 'length')],
      dtype='object')
```

but below not. (although just switch the position `idx` and `midx`)

```python
>>> midx ^ idx
Traceback (most recent call last):
...
TypeError: Expected tuple, got int
>>>
```

Reply:
@ueshin @HyukjinKwon 
I think there is some issue when we're handling `Index ^ MultiIndex`

i'm going to brief this issue with simple example, 

Here, we have an `Index` & a `MultiIndex` like below.

```python
>>> idx
Int64Index([1, 2, 3, 4], dtype='int64')
>>> midx
MultiIndex([('koalas',  'speed'),
            ('koalas',  'speed'),
            ('koalas',  'speed'),
            (   'cow',  'speed'),
            (   'cow', 'weight'),
            (   'cow', 'length'),
            ('falcon',  'speed'),
            ('falcon', 'weight'),
            ('falcon', 'length')],
           )
```

in pandas, they will show 'symdiff' result like below:

```python
>>> idx ^ midx
Index([                   1,                    2,                    3,
                          4,    ('lama', 'speed'),     ('cow', 'speed'),
          ('cow', 'weight'),    ('cow', 'length'),  ('falcon', 'speed'),
       ('falcon', 'weight'), ('falcon', 'length')],
      dtype='object')
```

the problem is here,

because i think now we can't support Index like this, since Internal spark dataframe can only support single type for single column, not multiple type

So i though how about we would support result of `Index ^ MultiIndex` as a string type.

Below example shows an internal spark dataframe for `symdiff` which type is string , 1 ~ 4 are from `Index` which type is bigint,  and the others from `MultiIndex` which type is struct<string, string>.

```
>>> sdf_symdiff.show()
+-----------------+
|__index_level_0__|
+-----------------+
|                1|
|                2|
|                3|
|                4|
|  [koalas, speed]|
|  [koalas, speed]|
|  [koalas, speed]|
|     [cow, speed]|
|    [cow, weight]|
|    [cow, length]|
|  [falcon, speed]|
| [falcon, weight]|
| [falcon, length]|
+-----------------+
```

then the final result will be shown as:

```python 
Index(['1', '2', '3', '4', '[koalas, speed]', '[koalas, speed]',
       '[koalas, speed]', '[cow, speed]', '[cow, weight]', '[cow, length]',
       '[falcon, speed]', '[falcon, weight]', '[falcon, length]'],
      dtype='object')
```

of course this result is slightly different from pandas' , but i think it still makes sense.

and we can make '[]' to '()', If we don't care about the cost of `replace`.

Have any suggestion to solve this problem with a better way?

Reply:
Maybe we should just not support such cases for now.

Reply:
@ueshin it seems better for now 👍